### PR TITLE
Refactor hotkey handling

### PR DESCRIPTION
**Pull request by bvschaik**
Preparation for possibly configurable hotkeys.

### CODE REVIEW

**Comment by crudelios** [comment]
I'll have a look at this. I'm just about to implement a `joystick_mapping` class, so I'll have a look at this codebase to see what ideas you came up with to keep the implementation similar.

File: src/game/cheats.c [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        scenario_invasion_start_from_cheat();
+    } else {
+        data.is_cheating = 0;
+    }
+}
+
+void game_cheat_money(void)
+{
+    if (data.is_cheating) {
+        city_victory_force_win();
```
Comment (crudelios) [inline src/game/cheats.c:29]: You mixed up the money and victory cheat! Almost as good as Caesar firing you in the map editor haha :wink:

File: src/window/city.c [hunk 2]
```diff
@@ -99,22 +108,127 @@ static void draw_foreground_military(void)
...
+    window_invalidate();
+}
+
+static void cycle_legion(void)
+{
+    static int current_legion_id = 1;
+    if (window_is(WINDOW_CITY)) {
+        int legion_id = current_legion_id;
+        current_legion_id = 0;
+        for (int i = 1; i <= MAX_LEGIONS; i++) {
```
Comment (crudelios) [inline src/window/city.c:136]: I know this bug isn't directly caused by your changes, but I think it should be `MAX_FORMATIONS` here, not `MAX_LEGIONS`.

File: src/window/plain_message_dialog.c [hunk 3]
```diff
@@ -47,24 +47,27 @@ static void draw_foreground(void)
...
+}
+
+static void handle_input(const mouse *m, const hotkeys *h)
 {
     if (image_buttons_handle_mouse(mouse_in_dialog(m), 80, 80, buttons, 1, 0)) {
         return;
     }
-    if (input_go_back_requested()) {
-        window_plain_message_dialog_accept();
+    if (input_go_back_requested(m, h)) {
```
Comment (crudelios) [inline src/window/plain_message_dialog.c:60]: Why not just:

```
if (input_go_back_requested(m, h) || h->enter_pressed) {
    window_go_back();
}
```

File: src/game/cheats.c (hunk 1 above)
Comment (bvschaik) [reply src/game/cheats.c:29]: I had too much sun yesterday 😂

File: src/window/city.c (hunk 2 above)
Comment (bvschaik) [reply src/window/city.c:136]: True. I'll fix that directly on master though.

File: src/window/plain_message_dialog.c (hunk 3 above)
Comment (bvschaik) [reply src/window/plain_message_dialog.c:60]: Probably because I was quite tired after refactoring about 70 screens. I'll simplify this.

Comment (bvschaik) [reply src/window/plain_message_dialog.c:60]: Probably because I was quite tired after refactoring about 70 screens. I'll simplify this.

**Comment by bvschaik** [comment]
> I'll have a look at this. I'm just about to implement a `joystick_mapping` struct, so I'll have a look at this codebase to see what ideas you came up with to keep the implementation similar.

The basic idea is: `hotkeys.c` should translate key presses into "commands" which are in the `hotkeys` struct. The individual screens will process those commands in their `handle_input` callback (previously called `handle_mouse`). This way, the mapping from key to command can be changed easily later.

I'm not sure how the mapping will look like yet though, as I'm still unsure how to proceed with regards to different keyboard layouts. Should the mapping be for physical keys or virtual ones? And how should the user configure the hotkeys? Just by editing an .ini file or in-game? If in-game, how?

**Comment by crudelios** [comment]
> Should the mapping be for physical keys or virtual ones?

After thinking about this, I think mapping should be for physical keys. It's what makes sense. Users should expect the same key to work regardless of the layout.

> And how should the user configure the hotkeys? Just by editing an .ini file or in-game? If in-game, how?

We could start with a simple ini file then implement a configuration window later. I'll be trying to create a configuration window for joystick so I can adapt it for regular hotkeys.

**Pull request merged**